README.md
# ansi2
[![Build Status](https://travis-ci.org/elidoran/node-ansi2.svg?branch=master)](https://travis-ci.org/elidoran/node-ansi2)
[![Dependency Status](https://gemnasium.com/elidoran/node-ansi2.png)](https://gemnasium.com/elidoran/node-ansi2)
[![npm version](https://badge.fury.io/js/ansi2.svg)](http://badge.fury.io/js/ansi2)
[![Coverage Status](https://coveralls.io/repos/github/elidoran/node-ansi2/badge.svg?branch=master)](https://coveralls.io/github/elidoran/node-ansi2?branch=master)

ANSI stream control of position, style, and color via fluent API.

ansi2 sends ANSI escape codes to a writable stream to:

1. move the cursor around
2. clear portions of the screen, or all of it
3. use standard ANSI colors and RGB colors for both foreground and background
4. accepts CSS style colors
5. show/hide the cursor
6. emit a beep sound


## Install

```sh
npm install --save ansi2
```


## Table of Contents

1. [Usage: Base](#usage-base)
2. [Usage: Colors](#usage-colors)
3. [Usage: Styles](#usage-styles)
4. [Usage: Position](#usage-position)


## Usage: Base

```javascript
// wrap a writable stream
var ansi = require('ansi2')(process.stdout)

// write a string to stream.
// it's affected by the current color/style/position.
// default styling + color:
ansi.write('boring default text')

// reset the foreground, background, and styling.
ansi.reset()

// Or, reset only the foreground:
ansi.resetFg()

// Or, reset only the background:
ansi.resetBg()

// all functions are chainable:
ansi.reset().blue().blackBg().write('blah').reset().write('\n')

// convenience method helps reset before writing a newline
// to avoid repeatedly writing: .reset().write('\n')
ansi.bold().blue('some bold blue text').newline()
    .underline().green('some underlined green text').newline()
```


## Usage: Colors

Available colors as functions:

1. **black**
2. **red**
3. **green**
4. **yellow**
5. **blue**
6. **magenta**
7. **cyan**
8. **white**
9. **grey/gray**

Use the "bright" versions of those by adding suffix "Bright" to the function name.

Apply the color to the background by adding suffix "Bg" to the function name.

Apply both suffixes for a bright background (add suffix "BrightBg").

Other color functions accept RGB values and CSS style hex codes:

1. `rgb(r, g, b)` - Also has a background version as `rgbBg(r, g, b)`
2. `hex(string)` - Also has a background version as `hexBg(string)`. May have a hash symbol prefix, or not.

Color functions called with an argument will change the color and then write the string.

```javascript
ansi.blue().write('blue text')
// Or, use a shorthand, provide the text to the color function:
ansi.blue('blue text')

// still using blue for foreground.
// change the background now.
// standard colors have a "background" version
// by adding 'Bg' to their function name.
ansi.blackBg().write('blue on black text')


// or specify colors via RGB or hex values.
// all these change the color to red.
ansi.rgb(255, 0, 0)
ansi.hex('FF0000')
ansi.hex('#FF0000')
```


## Usage: Styles

Available styles as functions:

1. **bold**
2. **italic**
3. **underline**
4. **inverse**

Reset the style by adding suffix "Reset" to the function name.

Style functions called with an argument will change the style and then write the string.

```javascript
ansi.bold().write('bold text')
// Or, use a shorthand, provide the text to the color function:
ansi.bold('bold text')

ansi.italic()
// then, undo the bold styling only, retaining italic.
ansi.boldReset()
```


## Usage: Position

Available position functions:

1. **up** - move cursor up one line, or, pass a number argument to specify line count
2. **down** - move cursor down one line, or, pass a number argument to specify line count
3. **forward** - move cursor right one column, or, pass a number argument to specify column count
4. **back** - move cursor left one column, or, pass a number argument to specify column count
5. **next** - move cursor to the beginning (left) one line down, or, pass a number argument to specify line count
6. **previous** - move cursor to the beginning (left) one line up, or, pass a number argument to specify line count
7. **left** - move cursor to the beginning of the current line, or, pass a number argument to specify the column to move to
8. **clearBelow** - clear all screen data below current cursor position
9. **clearAbove** - clear all screen data above current cursor position (not history)
10. **clear** - clear entire screen (not history)
11. **clearAll** - clear entire screen and all scroll history
12. **eraseBefore** - erase line content to the left of the cursor position
13. **eraseAfter** - erase line content to the right of the cursor position
14. **eraseLine** - erase entire current line
15. **scrollUp** - scroll up one line, or, specify a number argument to specify how many lines to scroll
16. **scrollDown** - scroll down one line, or, specify a number argument to specify how many lines to scroll
17. **save** - save current cursor position
18. **restore** - restore current cursor position
19. **position** - get the current cursor position
20. **hide** - hide the cursor
21. **show** - show the cursor
22. **restartLine** - convenience function to move to the beginning of the line and erase all line content. specify a number argument and it will go to that column and then erase content after that position.
23. **goto** - convenience function to move to a specific line and column on the screen (both default to 0)

```javascript
// example using default, this will move up one line:
ansi.up()

// example specifying amount to move, this moves up 5 lines:
ansi.up(5)

// moves to the top left:
ansi.goto()

// moves to line 5 column 7:
ansi.goto(7, 5) // think (x, y)

// erase current line and then write new content:
ansi.restartLine().write('new text overwriting current line')

// or, restart back at a certain column to overwrite only the later part:
ansi.write('label: some text')        // shows whole string
    .restartLine(7).write('new text') // shows: label: new text
```


## [MIT License](LICENSE)


CHANGES.md
## 1.0.0 - Released 2017/06/17

1. add linting via eslint with a config file in build/
2. tweak code a bit for lint
3. rename script "test-all" to "tests"
4. eliminate travis-only coverage script and use one for both local and travis


## 0.2.0 - Released 2017/06/10

1. remove newline counting thing to put into a separate package
2. remove closures cuz they make no noticeable performance difference.
3. change styles tracking internals from an "is" object to individual root props
4. add benchmarking versus package `ansi` (also tested some alternate implementations)
5. add `newline()` convenience method


## 0.1.0 - Released 2017/06/08

1. initial working version with 100% coverage


test/test.js
var assert  = require('assert')
var Emitter = require('events').EventEmitter

var strung = require('strung')

var build = require('../lib/index.js')

var fake = function() {
  var result = new Emitter
  result.chunks = []
  result.write = function(data) { result.chunks.push(data) }
  return result
}

describe('test ansi2', function() {

  it('should build', function() {
    assert(build(fake()))
  })

  it('should cache it and return that cached value', function() {
    var stream = fake()
    var ansi = build(stream)
    assert(ansi)
    assert.equal(stream._ansi2, ansi)
    assert.equal(build(stream), ansi)
  })

  it('should build Ansi w/out new', function() {
    assert(build.Ansi(fake()))
  })

  it('should error if not provided an object', function() {
    assert.throws(function() {
      new build.Ansi('string')
    })
  })

  it('should error if not provided an object with a write() function', function() {
    assert.throws(function() {
      new build.Ansi({})
    })
  })

  it('should be enabled (true) based on isTTY w/out options', function() {
    var stream   = fake()
    stream.isTTY = true
    var ansi     = build(stream)
    assert(ansi)
    assert.equal(ansi.enabled, true)
  })

  it('should be enabled (false) based on isTTY w/out options', function() {
    var stream   = fake()
    stream.isTTY = false
    var ansi    = build(stream)
    assert(ansi)
    assert.equal(ansi.enabled, false)
  })

  it('should build w/option enabled=true', function() {
    var stream  = fake()
    var options = { enabled: true }
    var ansi    = build(stream, options)
    assert(ansi)
    assert.equal(ansi.enabled, true)
  })

  it('should build w/option enabled=false', function() {
    var options = { enabled: false }
    var ansi = build(fake(), options)
    assert(ansi)
    assert.equal(ansi.enabled, false)
  })

  it('should build and start in buffer mode w/option buffering=true', function() {
    var options = { buffering: true }
    var ansi = build(fake(), options)
    assert(ansi)
    assert.equal(ansi.write, ansi.bufferWrite)
  })

  it('should build and start in write mode w/option buffering=false', function() {
    var options = { buffering: false }
    var ansi = build(fake(), options)
    assert(ansi)
    assert.equal(ansi.write, ansi.sendWrite)
  })

  it('should not override write if it already has been', function() {
    function fn() {}
    var stream = fake()
    var original = stream.write
    stream.realWrite = fn
    var ansi = build(stream)
    assert(ansi)
    assert.equal(stream.realWrite, fn)
    assert.equal(stream.write, original)
  })

  it('should buffer content', function() {
    var stream = fake()
    var ansi = build(stream)
    ansi.buffer()
    ansi.write('one')
    ansi.write('two').write('three')
    ansi.flush()
    assert.deepEqual(stream.chunks, ['one', 'two', 'three'])
  })

  it('should reset and write a newline from newline()', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.bold().italic().underline().inverse().blue()
    target.string = ''
    ansi.newline()
    assert.equal(target.string, '\x1b[0m\n')
  })

  it('should beep...', function() {
    build(fake(), {enabled:true}).beep()
  })

  it('should NOT beep when disabled', function() {
    build(fake(), {enabled:false}).beep()
  })

  it('should move to the far left and erase line', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.write('testing')
    ansi.write(' testing2')
    ansi.restartLine().write('retest it')
    assert.equal(target.string, 'testing testing2\x1b[G\x1b[0Kretest it')
  })

  it('should goto(1, 2)', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.goto(1, 2)
    assert.equal(target.string, '\x1b[2;1H')
  })

  it('should not goto() when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.goto(1, 2)
    assert.equal(target.string, null)
  })

  it('should back()', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.back()
    assert.equal(target.string, '\x1b[D')
  })

  it('should back(2)', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.back(2)
    assert.equal(target.string, '\x1b[2D')
  })

  it('should not back() when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.back()
    assert.equal(target.string, null)
  })


  it('should reset', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.isBold = true
    ansi.isItalic = true
    ansi.isUnderline = true
    ansi.isInverse = true
    ansi.fgCurrent = 123
    ansi.bgCurrent = 123
    ansi.reset()
    assert.equal(target.string, '\x1b[0m')
    assert.equal(ansi.isBold, false)
    assert.equal(ansi.isItalic, false)
    assert.equal(ansi.isUnderline, false)
    assert.equal(ansi.isInverse, false)
    assert.equal(ansi.fgCurrent, 39)
    assert.equal(ansi.bgCurrent, 49)
  })

  it('should reset() w/out writing reset when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.isBold = true
    ansi.isItalic = true
    ansi.isUnderline = true
    ansi.isInverse = true
    ansi.fgCurrent = 123
    ansi.bgCurrent = 123
    ansi.reset()
    assert.equal(target.string, null)
    assert.equal(ansi.isBold, false)
    assert.equal(ansi.isItalic, false)
    assert.equal(ansi.isUnderline, false)
    assert.equal(ansi.isInverse, false)
    assert.equal(ansi.fgCurrent, 39)
    assert.equal(ansi.bgCurrent, 49)
  })

  it('should resetFg()', function() {
    var ansi = build(fake(), {enabled:true})
    ansi.fgCurrent = 123
    ansi.resetFg()
    assert.equal(ansi.fgCurrent, 39)
  })

  it('should resetBg()', function() {
    var ansi = build(fake(), {enabled:true})
    ansi.bgCurrent = 123
    ansi.resetBg()
    assert.equal(ansi.bgCurrent, 49)
  })

  it('should use RGB value (black) in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.rgb(0, 0, 0)
    assert.equal(target.string, '\x1b[38;5;16m')
  })

  it('should use RGB value (white) in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.rgb(255, 255, 255)
    assert.equal(target.string, '\x1b[38;5;231m')
  })

  it('should use RGB value (black) in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.rgbBg(0, 0, 0)
    assert.equal(target.string, '\x1b[48;5;16m')
  })

  it('should use RGB value (white) in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.rgbBg(255, 255, 255)
    assert.equal(target.string, '\x1b[48;5;231m')
  })

  it('should use hex color 000000 in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hex('000000')
    assert.equal(target.string, '\x1b[38;5;16m')
  })

  it('should use hex color #000000 in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hex('#000000')
    assert.equal(target.string, '\x1b[38;5;16m')
  })

  it('should use hex color FFFFFF in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hex('FFFFFF')
    assert.equal(target.string, '\x1b[38;5;231m')
  })

  it('should use hex color #FFFFFF in foreground', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hex('#FFFFFF')
    assert.equal(target.string, '\x1b[38;5;231m')
  })

  it('should use hex color 000000 in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hexBg('000000')
    assert.equal(target.string, '\x1b[48;5;16m')
  })

  it('should use hex color #000000 in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hexBg('#000000')
    assert.equal(target.string, '\x1b[48;5;16m')
  })

  it('should use hex color FFFFFF in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hexBg('FFFFFF')
    assert.equal(target.string, '\x1b[48;5;231m')
  })

  it('should use hex color #FFFFFF in background', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.hexBg('#FFFFFF')
    assert.equal(target.string, '\x1b[48;5;231m')
  })

  it('should skip setting the same foreground color', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.blue()
    ansi.blue()
    ansi.blue()
    assert.equal(target.string, '\x1b[34m')
  })

  it('should skip setting the same background color', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.blueBg()
    ansi.blueBg()
    ansi.blueBg()
    assert.equal(target.string, '\x1b[44m')
  })

  it('should skip writing the foreground color when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.blue()
    ansi.blue()
    ansi.blue()
    assert.equal(target.string, null)
  })

  it('should skip writing the background color when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.blueBg()
    ansi.blueBg()
    ansi.blueBg()
    assert.equal(target.string, null)
  })

  it('should write data provided to a foreground color function', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.blue('test')
    assert.equal(target.string, '\x1b[34mtest')
  })

  it('should write data provided to a background color function', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.blueBg('test')
    assert.equal(target.string, '\x1b[44mtest')
  })

  it('should bold() and boldReset()', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.bold().write('test').boldReset()
    assert.equal(target.string, '\x1b[1mtest\x1b[22m')
  })

  it('should bold(\'test\') and boldReset()', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.bold('test').boldReset()
    assert.equal(target.string, '\x1b[1mtest\x1b[22m')
  })

  it('should not output bold() when already bolded', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.isBold = true
    ansi.bold('test')
    assert.equal(target.string, 'test')
  })

  it('should not output boldReset() when not bolded', function() {
    var target = strung()
    var ansi = build(target, {enabled:true})
    ansi.boldReset()
    assert.equal(target.string, null)
  })

  it('should not output bold() when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.bold('test')
    assert.equal(target.string, 'test')
  })

  it('should not output boldReset() when disabled', function() {
    var target = strung()
    var ansi = build(target, {enabled:false})
    ansi.isBold = true
    ansi.boldReset()
    assert.equal(target.string, null)
  })

})


lib/index.js
// 'ansi2' is adapted from package 'ansi'. thank you.

var PREFIX       = '\x1b['
var COLOR_SUFFIX = 'm'
var RESET        = PREFIX + '0' + COLOR_SUFFIX

module.exports      = build
module.exports.Ansi = Ansi // export it as a sub-property

// return one we already made or make a new one
function build(writable, options) {
  return writable && writable._ansi2 || (writable._ansi2 = new Ansi(writable, options))
}

function Ansi(writable, options) {

  if (!(this instanceof Ansi)) {
    return new Ansi(writable, options)
  }

  if (typeof writable != 'object' || typeof writable.write != 'function') {
    throw new Error('writable required')
  }

  this.writable  = writable
  this.enabled   = options ? (options.enabled === true) : writable.isTTY === true
  this.write     = (options && options.buffering) ? this.bufferWrite : this.sendWrite
  this._buffer   = []
  this.fgCurrent = 39
  this.bgCurrent = 49
  this.isBold    = false
  this.isItalic  = false
  this.isInverse = false
  this.isUnderline = false
}


Ansi.prototype = {

  // one of these two are assigned to this.write to handle buffering/writing.
  sendWrite  : function(data) { this.writable.write(data) ; return this },
  bufferWrite: function(data) { this._buffer.push(data)   ; return this },

  // switch to buffer mode by changing write() from sendWrite() to bufferWrite()
  buffer: function () { this.write = this.bufferWrite ; return this },

  // switch to non-buffer mode by changing write() from bufferWrite() to sendWrite().
  // and sends buffered content.
  flush: function () {
    var i
    this.write = this.sendWrite
    for (i = 0; i < this._buffer.length; i++) this.write(this._buffer[i])
    this._buffer.length = 0
    return this
  },

  restartLine: function(pos) { return this.left(pos || 0).eraseAfter() },

  newline: function newline() { return this.reset().write('\n') },

  beep: function beep() {
    if (this.enabled) this.write('\x07')
    return this
  },

  goto: function goto(x, y) {
    if (this.enabled) this.write(PREFIX + (y | 0) + ';' + (x | 0) + 'H')
    return this
  },

  reset: function reset() {
    if (this.enabled) this.write(RESET)
    this.isBold      = false
    this.isItalic    = false
    this.isUnderline = false
    this.isInverse   = false
    this.fgCurrent    = 39
    this.bgCurrent    = 49
    return this
  },

  resetFg: function resetFg() { return this._setFgColor('39', '\x1b[39m') },
  resetBg: function resetBg() { return this._setBgColor('49', '\x1b[49m') },

  rgb: function rgb(r, g, b) {
    return this._setFgColor('38;5;' + this._rgb(r, g, b), null)
  },

  rgbBg: function rgb(r, g, b) {
    return this._setBgColor('48;5;' + this._rgb(r, g, b), null)
  },

  // convert trio of 8 bit color values to single ansi color code.
  _rgb: function (r, g, b) {
    return 16
      + (Math.round(r / 255 * 5) * 36)
      + (Math.round(g / 255 * 5) *  6)
      + (Math.round(b / 255 * 5) *  1)
  },

  hex: function (color) {
    var colors = this._hex(color)
    return this.rgb(colors[0], colors[1], colors[2])
  },

  hexBg: function (color) {
    var colors = this._hex(color)
    return this.rgbBg(colors[0], colors[1], colors[2])
  },

  _hex: function (color) { // accepts CSS style color code
    var s = color[0] === '#' ? 1 : 0  // start index
    return [ // eslint-disable-next-line computed-property-spacing
      parseInt(color[s    ], 16) * 16 + parseInt(color[s + 1], 16),
      parseInt(color[s + 2], 16) * 16 + parseInt(color[s + 3], 16),
      parseInt(color[s + 4], 16) * 16 + parseInt(color[s + 5], 16),
    ]
  },

  _setFgColor: function setFgColor(code, string) {
    if (this.fgCurrent !== code) {
      if (this.enabled) this.write(string || PREFIX + code + COLOR_SUFFIX)
      this.fgCurrent = code
    }
    return this
  },

  _setBgColor: function setBgColor(code, string) {
    if (this.bgCurrent !== code) {
      if (this.enabled) this.write(string || PREFIX + code + COLOR_SUFFIX)
      this.bgCurrent = code
    }
    return this
  },
};


[ // add position functions
  ['up'         , 'A'],
  ['down'       , 'B'],
  ['forward'    , 'C'],
  ['back'       , 'D'],
  ['next'       , 'E'],
  ['previous'   , 'F'],
  ['left'       , 'G'],
  ['clearBelow' , '0J'],
  ['clearAbove' , '1J'],
  ['clear'      , '2J'],
  ['clearAll'   , '3J'],
  ['eraseBefore', '1K'],
  ['eraseAfter' , '0K'],
  ['eraseLine'  , '2K'],
  ['scrollUp'   , 'S'],
  ['scrollDown' , 'T'],
  ['save'       , 's'],
  ['restore'    , 'u'],
  ['position'   , '6n'],
  ['hide'       , '?25l'],
  ['show'       , '?25h'],
].forEach(function (values) {

  var code = values[1]

  Ansi.prototype[values[0]] = function(first) {
    var output, i

    if (this.enabled) {

      if (first) {
        output = PREFIX + Math.round(first)

        // none of these take multiple arguments...
        // istanbul ignore next
        for (i = 1; i < arguments.length; i++) {
          output += ';' + Math.round(arguments[i])
        }
      }

      else {
        output = PREFIX
      }

      this.write(output + code)
    }

    return this
  }
});


[ // add style functions (name, code, reset)
  ['bold'     , 1, 22],
  ['italic'   , 3, 23],
  ['underline', 4, 24],
  ['inverse'  , 7, 27],
].forEach(function (values) {

  var name  = 'is' + values[0][0].toUpperCase() + values[0].substring(1)
  var code  = PREFIX + values[1] + COLOR_SUFFIX
  var reset = PREFIX + values[2] + COLOR_SUFFIX

  Ansi.prototype[values[0]] = function(data) {
    if (!this[name]) {
      if (this.enabled) this.write(code)
      this[name] = true
    }

    if (data) this.write(data)

    return this
  }

  Ansi.prototype[values[0] + 'Reset'] = function() {
    if (this[name]) {
      if (this.enabled) this.write(reset)
      this[name] = false
    }

    return this
  }
});


[ // add color functions  (color, foreground code)
  ['black', 30],
  ['red'    , 31],
  ['green'  , 32],
  ['yellow' , 33],
  ['blue'   , 34],
  ['magenta', 35],
  ['cyan'   , 36],
  ['white'  , 37],
  ['grey'   , 90],
  ['gray'   , 90],
  ['blackBright'  , 90],
  ['redBright'    , 91],
  ['greenBright'  , 92],
  ['yellowBright' , 93],
  ['blueBright'   , 94],
  ['magentaBright', 95],
  ['cyanBright'   , 96],
  ['whiteBright'  , 97],
].forEach(function(values) {

  var fgCode   = values[1]
  var bgCode   = values[1] + 10
  var fgString = PREFIX + fgCode + COLOR_SUFFIX
  var bgString = PREFIX + bgCode + COLOR_SUFFIX

  Ansi.prototype[values[0]] = function(data) {
    this._setFgColor(fgCode, fgString)
    if (data) this.write(data)
    return this
  }

  Ansi.prototype[values[0] + 'Bg'] = function(data) {
    this._setBgColor(bgCode, bgString)
    if (data) this.write(data)
    return this
  }
})
